doc/mgmt/gnmi/gNMI_Subscription_for_YangData.md

```diff
@@ -0,0 +1,1183 @@
+  - [2.14 Handling YANG Paths without Module Prefix](#214-handling-yang-paths-without-module-prefix)
+  - [2.15 Wildcard Keys for Non-DB paths](#215-wildcard-keys-for-non-db-paths)
+- [3 Design Details](#3-design-details)
+  - [3.1 Translib APIs](#31-translib-apis)
+    - [3.1.1 SubscribeSession](#311-subscribesession)
+    - [3.1.2 IsSubscribeSupported](#312-issubscribesupported)
+    - [3.1.3 Subscribe](#313-subscribe)
+    - [3.1.4 Stream](#314-stream)
+  - [3.2 App Interface](#32-app-interface)
+    - [3.2.1 translateSubscribe](#321-translatesubscribe)
+    - [3.2.2 processSubscribe](#322-processsubscribe)
+    - [3.2.3 ProcessOnChange](#323-processonchange)
+  - [3.3 CommonApp module](#33-commonapp-module)
+    - [3.3.1 CommonApp's TranslateSubscribe](#331-commonapps-translatesubscribe)
+    - [3.3.2 CommonApp's ProcessSubscribe](#332-commonapps-processsubscribe)
+  - [3.4 Transformer](#34-transformer)
+    - [3.4.1 Subtree Subscribe transformer](#341-subtree-subscribe-transformer)
+    - [3.4.2 Path transformer](#342-path-transformer)
+    - [3.4.3 Key transformer](#343-key-transformer)
+    - [3.4.4 Table transformer](#344-table-transformer)
+  - [3.5 Path validator](#35-path-validator)
+- [4 User Interface](#4-user-interface)
+- [5 Serviceability and Debug](#5-serviceability-and-debug)
+  - [5.1 Finding Subscription Preferences for a YANG Path](#51-finding-subscription-preferences-for-a-yang-path)
+- [6 Scale Considerations](#6-scale-considerations)
+- [7 Limitations](#7-limitations)
+- [8 Unit Tests](#8-unit-tests)
+
+## Revision History
+
+| Rev | Date        |       Author       | Change Description                                                            |
+|-----|-------------|--------------------|-------------------------------------------------------------------------------|
+| 0.1 | 03/02/2023  | Sachin Holla       | Initial draft                                                                 |
+| 0.2 | 03/14/2023  | Balachandar Mani   | Added the CommonApp's interface, Transformer and Path validator details |
+
+## Definition and Abbreviation
+
+| **Term**       | **Meaning**                         |
+|----------------|-------------------------------------|
+| gNMI           | gRPC Network Management Interface   |
+| gRPC           | gRPC Remote Procedure Call (it is a recursive acronym!!)   |
+| RPC            | Remote Procedure Call.. In this document it mostly refers to the operations defined by gNMI specification |
+| YGOT           | [YANG Go Tools](https://github.com/openconfig/ygot); a Golang library to process YANG data |
+
+## 1 Feature Overview
+
+### 1.1 Introduction
+
+SONiC Telemetry service suports gNMI Get, Set and Subscribe RPCs for DB paths and sonic-yang based paths.
+It also suppots Get and Set for OpenConfig and IETF yang based paths that are part of **sonic-mgmt-common** repository.
+This design document describes proposed enhancements to support gNMI Subscribe RPC for such YANG paths.
```

Comment:
>YANG paths [](http://example.com/codeflow?start=92&length=10)

How about sonic-yang paths? If out of scope of this HLD, do you evaluate the possibility and effort needed?

Reply:
I see some of the requirements are useful for sonic-yang paths, and for sonic-db paths. For example TARGET_DEFINED/POLL and ONCE/Wildcard Keys.

It would be nice if the requirement could be implemented onto all the possible path types.

Reply:
Translib can handle sonic-yang paths. We planning to implement in a future release.

sonic-db paths cannot be be handled in translib. You need to enhance the DbClient implementation to handle all other subscription modes. This HLD is applicable for TranslClient only.